src/containers/Freelancer/logic.js

```diff
@@ -0,0 +1,78 @@
+// library dependencies
+import { createLogic } from 'redux-logic';
+
+// custom actions
+import { actionTypes, actions } from './actions';
+
+// custom utils
+import API from '../../utils/Api';
+import Debug from '../../utils/Debug';
+
+
+export const freelancerQueryLogic = createLogic(
+  {
+    type: actionTypes.FREELANCER_QUERY,
+    /* let's prevent empty requests */
+    validate: ({ action }, allow, reject) => {
+      if ( action.type === 'FREELANCER_QUERY' && action.payload) {
+        Debug.log(['[action:validate:allow] FREELANCER_QUERY'], action.payload)
+        allow(action);
+      } else {  /* empty request, silently reject */
+        reject();
+      }
+    },
+    process: ({ action, getState }, dispatch, done) => {
+      const state = getState();
+      Debug.log('calling] FREELANCER_QUERY', action.payload);
+      API.Axios('POST', '/graphql', API.generateFreelancerInfoQuery(action.payload), state.Login.jwt)
+      .then ( data => {
+        if ( data ) {
+          Debug.log('[action:process:freelancer_info] FREELANCER_QUERY', data);
+          dispatch(actions.freelancerInfo(data));
+        } else {
+          throw(new Error('Request failed.'));
+        }
+      })
+      .catch( error => {
+        throw(new Error(error));
+      })
+      .then ( () => {
+        done();
+      });
+    }
+  }
+);
+
+export const freelancerInfoLogic = createLogic(
+  {
+    type: actionTypes.FREELANCER_INFO,
+    /* let's prevent empty requests */
+    validate: ({ action }, allow, reject) => {
+      if ( action.type === 'FREELANCER_INFO' && action.payload) {
+        Debug.log(['[action:validate:allow] FREELANCER_INFO'], action.payload)
+        allow(action);
+      } else {  /* empty request, silently reject */
+        reject();
+      }
+    },
+    process: ({ action, getState }, dispatch, done) => {
+      const state = getState();
+      Debug.log('calling] FREELANCER_INFO', action.payload);
+      API.Axios('POST', '/graphql', API.generateFreelancerInfoQuery(action.payload), state.Login.jwt)
+      .then ( data => {
+        console.log(data);
+      })
+      .catch( error => {
+        throw(new Error(error));
+      })
+      .then ( () => {
+        done();
+      });
```

Comment:
The 2nd API call in the ``process`` hook of the logic for action ``FREELANCER_INFO`` is redundant. You can remove it.


I didn't find Redux Logic intuitive at first either, but after the ``validate`` hook runs, logic lets the action go through to the reducer. the ``process`` hook is run **after**. 

So in effect your first action (in this case ``FREELANCER_QUERY``) is the initiator of the flow of retrieving data, but and the API call is a side effect.

The actions as far as redux is concerned look like this:
1. FREELANCER_QUERY => 
2. intercepted by VALIDATE => 
3. OK?  => (*)
4. REDUCER => 
5. END
but at the point of (*) Logic passes to its own ``process`` hook where you can do async stuff and dispatch other actions

hence...

process => API => result => FREELANCER_INFO(result).

Technically, you don't even need the logic for FREELANCER_INFO action, but it's good to ``validate`` it and make sure the payload is passed.